lib/semantics/expression.cc

```diff
@@ -1470,9 +1474,46 @@ static MaybeExpr AnalyzeExpr(ExpressionAnalysisContext &context,
       }
       unavailable.insert(symbol->name());
       if (MaybeExpr value{AnalyzeExpr(context, expr)}) {
-        // TODO pmk: C7104, C7105 check that pointer components are
-        // being initialized with data/procedure designators appropriately
-        result.Add(*symbol, std::move(*value));
+        bool isNULL{std::holds_alternative<NullPointer>(value->u)};
+        if (symbol->has<semantics::ProcEntityDetails>()) {
+          CHECK(symbol->attrs().test(semantics::Attr::POINTER));
+          if (!isNULL) {
+            // TODO C7104: check that procedure pointer components are
+            // being initialized with compatible procedure designators
+            context.Say(expr.source,
+                "TODO: non-null procedure pointer component value not implemented yet"_err_en_US);
+          }
+        } else {
+          CHECK(symbol->has<semantics::ObjectEntityDetails>());
+          // C1594(4)
+          if (!isNULL) {
+            const auto &innermost{context.context().FindScope(expr.source)};
+            if (const auto *pureFunc{
+                    semantics::FindPureFunctionContaining(&innermost)}) {
+              if (semantics::IsOrHasPointerComponent(*symbol) &&
+                  semantics::IsExternallyVisibleObject(*value, *pureFunc)) {
+                context.Say(expr.source,
+                    "Externally visible object must not be associated with a pointer in a PURE function"_err_en_US);
```

Comment:
Would this message be more helpful if it referred to the symbol?

Reply:
Yes, when the component is a pointer; maybe not when the component contains a nested pointer component several levels down.  From the expression, it's not always possible to dig out a clear symbol for the globally visible object -- it can be USE associated with renaming.  But I can try to do better on both, maybe by extending those predicates to return symbol pointers instead of simple truth values.

Reply:
Done.